File: Beacon/Signaling.m
//
//  Signaling.m
//  Beacon
//
//  Created by Pavel Skaldin on 2/24/19.
//  Copyright © 2019 Pavel Skaldin. All rights reserved.
//

#import "Signaling.h"
#import <Beacon/Beacon-Swift.h>

void _BeaconEmit(id object, NSArray<Beacon*>* beacons, NSDictionary* userInfo, NSString* file, NSInteger line, NSString* function) {
    Signal *signal = nil;
    if (object == nil) {
        signal = [[ContextSignal alloc] initWithStack:[NSThread callStackSymbols]];
    }
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"
    else if ([object respondsToSelector:@selector(beaconSignal)]) {
        signal = [object performSelector:@selector(beaconSignal)];
    }
#pragma clang diagnostic pop
    else {
        signal = [[WrapperSignal alloc] init:object];
    }
    [signal emitOn:(beacons != nil) ? beacons : @[[Beacon shared]]
          userInfo:userInfo
          fileName:file
              line:line
      functionName:function];
}

void _BeaconEmitError(NSError* error, NSArray<Beacon*>* beacons, NSDictionary* userInfo, NSString* file, NSInteger line, NSString* function) {
    ErrorSignal *signal = [[ErrorSignal alloc] initWithError:error stack: [NSThread callStackSymbols]];
    [signal emitOn:(beacons != nil) ? beacons : @[[Beacon shared]]
          userInfo:userInfo
          fileName:file
              line:line
      functionName:function];
}

void _BeaconEmitStackTrace(NSArray<NSString *>* stackTrace, NSArray<Beacon*>* beacons, NSDictionary* userInfo, NSString* file, NSInteger line, NSString* function) {
    StackTraceSignal *signal = [[StackTraceSignal alloc] initWithStackTrace:stackTrace];
    [signal emitOn:(beacons != nil) ? beacons : @[[Beacon shared]]
          userInfo:userInfo
          fileName:file
              line:line
      functionName:function];
}

File: Beacon/Signaling.swift
//
//  Signaling.swift
//  Beacon
//
//  Created by Pavel Skaldin on 1/29/19.
//  Copyright © 2019 Pavel Skaldin. All rights reserved.
//

import Foundation

/**
 I describe a signaling protocol.
 
 Conforming objects become `emit()`able, by returning an instance of appropriate `Signal`.
 
 For example:
 ````
 extension URLRequest : Signaling {
    var beaconSignal: Signal {
        return URLRequestSignal(self)
    }
 }
 
 class URLRequestSignal: Signal {
    var urlRequest: URLRequest
    init(_ aRequest: URLRequest) {
        urlRequest = aRequest
        super.init()
    }
 }
 ````
 
 In similar manner, `WrapperSignal` makes `emit()`able any class or construct...
 
 - See Also: `WrapperSignal`
 
 */
@objc public protocol Signaling {
    var beaconSignal: Signal { get }
}

/**
 Emits a signaling object.
 This is a convenience method to emit a signal associated with a `Signaling` object.
 To be consistent, it is preferred that all logging is done via various `emit()` methods,
 rather than directly interfacing with instances of `Signal` and `SignalLogger`.
 */
public func emit(_ value: Signaling, on beacon: Beacon = Beacon.shared, userInfo: [AnyHashable : Any]? = nil, fileName: String = #file, line: Int = #line, functionName: String = #function) {
    value.beaconSignal.emit(on: [beacon],
                            userInfo: userInfo,
                            fileName: fileName,
                            line: line,
                            functionName: functionName)
}

/**
 Emits a signaling object.
 This is a convenience method to emit a signal associated with a `Signaling` object.
 To be consistent, it is preferred that all logging is done via various `emit()` methods,
 rather than directly interfacing with instances of `Signal` and `SignalLogger`.
 */
public func emit(_ value: Signaling, on beacons: [Beacon], userInfo: [AnyHashable : Any]? = nil, fileName: String = #file, line: Int = #line, functionName: String = #function) {
    value.beaconSignal.emit(on: beacons,
                            userInfo: userInfo,
                            fileName: fileName,
                            line: line,
                            functionName: functionName)
}
